skrifa/src/glyph_name.rs

```diff
@@ -0,0 +1,394 @@
+    }
+
+    /// Returns the name for the given glyph identifier.
+    pub fn get(&self, glyph_id: GlyphId) -> Option<GlyphName> {
+        if glyph_id.to_u32() >= self.num_glyphs() {
+            return None;
+        }
+        let name = match &self.inner {
+            Inner::Post(post, _) => GlyphName::from_post(post, glyph_id),
+            Inner::Cff(cff, charset) => charset
+                .string_id(glyph_id)
+                .ok()
+                .and_then(|sid| GlyphName::from_cff_sid(cff, sid)),
+            _ => None,
+        };
+        Some(name.unwrap_or_else(|| GlyphName::synthesize(glyph_id)))
+    }
+
+    /// Returns an iterator yielding the identifier and name for all glyphs in
+    /// the font.
+    pub fn iter(&self) -> impl Iterator<Item = (GlyphId, GlyphName)> + 'a + Clone {
+        match &self.inner {
+            Inner::Post(post, n) => Iter::Post(0..*n, post.clone()),
+            Inner::Cff(cff, charset) => Iter::Cff(cff.clone(), charset.iter()),
+            Inner::Synthesized(n) => Iter::Synthesized(0..*n),
+        }
+    }
+}
+
+/// Specifies the chosen source for glyph names.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum GlyphNameSource {
+    /// Glyph names are sourced from the `post` table.
+    Post,
+    /// Glyph names are sourced from the `CFF` table.
+    Cff,
+    /// Glyph names are synthesized in the format `gidDDD` where `DDD` is
+    /// the glyph identifier in decimal.
+    Synthesized,
+}
+
+/// The name of a glyph.
+#[derive(Clone)]
+pub struct GlyphName {
+    name: [u8; MAX_GLYPH_NAME_LEN],
+    len: u8,
+    is_synthesized: bool,
+}
+
+impl GlyphName {
+    /// Returns the underlying name as a string.
+    pub fn as_str(&self) -> &str {
+        let bytes = &self.name[..self.len as usize];
+        core::str::from_utf8(bytes).unwrap_or_default()
+    }
+
+    /// Returns true if the glyph name was synthesized, i.e. not found in any
+    /// source.
+    pub fn is_synthesized(&self) -> bool {
+        self.is_synthesized
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Self {
+        let mut name = Self::default();
+        name.append(bytes);
+        name
+    }
+
+    fn from_post(post: &Post, glyph_id: GlyphId) -> Option<Self> {
+        glyph_id
+            .try_into()
+            .ok()
+            .and_then(|id| post.glyph_name(id))
+            .map(|s| s.as_bytes())
+            .map(Self::from_bytes)
+    }
+
+    fn from_cff_sid(cff: &Cff, sid: Sid) -> Option<Self> {
+        cff.string(sid)
+            .and_then(|s| core::str::from_utf8(s.bytes()).ok())
+            .map(|s| s.as_bytes())
+            .map(Self::from_bytes)
+    }
+
+    fn synthesize(glyph_id: GlyphId) -> Self {
+        use core::fmt::Write;
+        let mut name = Self {
+            is_synthesized: true,
+            ..Self::default()
+        };
+        let _ = write!(GlyphNameWrite(&mut name), "gid{}", glyph_id.to_u32());
+        name
+    }
+
+    fn append(&mut self, bytes: &[u8]) {
```

Comment:
do we ever expect this to be called on a non-empty name? I'm sure there's some rationale here but it seems like a funny API choice. I suppose it's just to work with `write!`? In that case I think it deserves a little expository comment.

Reply:
correct, the append functionality exists just for `write!` since we have no idea how many times `write_str` may be called for a given format. Added a comment explaining this.